log-warper-core/src/Log/Core/Class.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE DefaultSignatures          #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+-- | Monad type class for things that have logging ability.
+
+module Log.Core.Class
+       ( -- * Interface
+         MonadLogger (..)
+       , logR
+       , withR
+       , unliftLogAction
+
+         -- * Basic interface implementation
+       , ActionT (..)
+       , usingActionT
+       ) where
+
+import Control.Monad (void)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Class (MonadTrans (..))
+import Control.Monad.Trans.Except (ExceptT, mapExceptT, runExceptT)
+import Control.Monad.Trans.Identity (IdentityT (..), mapIdentityT)
+import Control.Monad.Trans.Maybe (MaybeT (..), mapMaybeT)
+import Control.Monad.Trans.Reader (ReaderT (..), ask, mapReaderT)
+import Control.Monad.Trans.State.Strict (StateT (..), get, mapStateT)
+import Data.Kind (Type)
+
+import Log.Core.Action (LogAction (..), cmap)
+
+{- | Type class for monads that have logging ability. In order to log things
+monad @m@ should be able to construct 'LogAction' for some logging record @r@.
+
+__Laws:__
+
+Some laws duplicate @lens@ laws. And maybe of the laws are redundant and can be
+proven from other laws.
+
+* __@Idempotency:@__ @askLogAction@ shouldn't mutate monadic context:
+
+@
+'askLogAction' >> 'askLogAction' ≡ 'askLogAction'
+@
+
+* __@Identity:@__ 'id' function shouldn't change anything:
+
+@
+'withLogAction' id ≡ id
+@
+
+* __@Composition:@__ composition of two 'withLogAction' is a single
+  'withLogAction' with function composition:
+
+@
+'withLogAction' f . 'withLogAction' g ≡ 'withLogAction' (f . g)
+@
+
+* __@PutGet:@__ 'askLogAction' after 'withLogAction' should return modified action.
+
+@
+'withLogAction' f 'askLogAction' ≡ 'fmap' f 'askLogAction'
+@
+-}
+class Monad m => MonadLogger (m :: Type -> Type) (r :: Type) | m -> r where
+    -- | Return 'LogAction' from current monadic context.
+    askLogAction :: m (LogAction m r)
+
+    default askLogAction :: (MonadTrans t, t n ~ m, MonadLogger n r)
+                         => m (LogAction m r)
+    askLogAction = do
```

Comment:
I'm not a fan of `default` method implementations. The problem is that sometimes one might forget to implement some method and there will be no warning.

Instead I would export `askLogActionLifted` as a separate function from the library, and the user could simply write `askLogAction = askLogActionLifted`, explicitly specifying the behavior she wants.

Reply:
@int-index Yeah, another problem with `default` implementations is that it doesn't work well with `{-#  MINIMAL #-}` pragma :disappointed: I like `default` implementations when instances can be one-liners. But in this case they can't, so probably having default function is better solution.